### PR TITLE
Use JDBI for selectors

### PR DESCRIPTION
**Pull request by oneonestar**
## Description
Relates to https://github.com/trinodb/trino-gateway/issues/229

(x) This is not user-visible or is docs only, and no release notes are required.

### CODE REVIEW

**Comment by oneonestar** [comment]
Before this PR, `testDeleteSelector()` a delete selector operator purged the cache for resource_groups table.
That's why `testReadSpecificDbResourceGroupCauseException` won't hit the cache.

After migrate selector to JDBI, `testDeleteSelector()` no longer purged the cache and `testReadSpecificDbResourceGroupCauseException` hit the cache then failed.
Idk why the cache hit across different jdbc url in activejdbc.. 

Use `-Dactivejdbc.log ` produce the following logs.
Before:
```
testDeleteSelector():
11:34:36.809 [main] DEBUG org.javalite.activejdbc.DB -- Opened connection: conn20: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA
11:34:36.809 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Attached connection named: default: to current thread: conn20: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA. Extra info: jdbc:h2:/tmp/h2db-1710588875777
11:34:36.809 [main] INFO org.javalite.activejdbc.LazyList -- {"sql":"SELECT * FROM selectors","params":[],"cache":"hit"}
11:34:36.810 [main] DEBUG org.javalite.activejdbc.DB -- Closed connection: conn20: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA
11:34:36.810 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Detached connection named: default from current thread: conn20: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA
11:34:36.812 [main] DEBUG org.javalite.activejdbc.DB -- Opened connection: conn21: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA
11:34:36.812 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Attached connection named: default: to current thread: conn21: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA. Extra info: jdbc:h2:/tmp/h2db-1710588875777
11:34:36.813 [main] INFO org.javalite.activejdbc.DB -- {"sql":"DELETE FROM selectors WHERE resource_group_id = 0 and priority = 0 and user_regex = 'data-platform-admin_updated' and source_regex = 'admin_updated' and query_type = 'query_type_updated' and client_tags = 'client_tag_updated' and selector_resource_estimate = 'estimate_updated'","params":[],"duration_millis":1}
11:34:36.813 [main] DEBUG org.javalite.activejdbc.cache.CacheManager -- Cache purged: table: selectors
11:34:36.813 [main] DEBUG org.javalite.activejdbc.cache.CacheManager -- Cache purged: table: resource_groups
11:34:36.814 [main] DEBUG org.javalite.activejdbc.DB -- Closed connection: conn21: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA
11:34:36.814 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Detached connection named: default from current thread: conn21: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA
11:34:36.816 [main] DEBUG org.javalite.activejdbc.DB -- Opened connection: conn22: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA
11:34:36.816 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Attached connection named: default: to current thread: conn22: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA. Extra info: jdbc:h2:/tmp/h2db-1710588875777
11:34:36.817 [main] INFO org.javalite.activejdbc.LazyList -- {"sql":"SELECT * FROM selectors","params":[],"duration_millis":1,"cache":"miss"}
11:34:36.817 [main] DEBUG org.javalite.activejdbc.DB -- Closed connection: conn22: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA
11:34:36.817 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Detached connection named: default from current thread: conn22: url=jdbc:h2:/tmp/h2db-1710588875777 user=SA

testReadSpecificDbResourceGroupCauseException():
11:34:36.978 [main] DEBUG org.javalite.activejdbc.DB -- Opened connection: conn34: url=jdbc:h2:/tmp/abcd user=SA
11:34:36.978 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Attached connection named: default: to current thread: conn34: url=jdbc:h2:/tmp/abcd user=SA. Extra info: jdbc:h2:/tmp/abcd
11:34:36.979 [main] DEBUG org.javalite.activejdbc.DB -- Closed connection: conn34: url=jdbc:h2:/tmp/abcd user=SA
11:34:36.980 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Detached connection named: default from current thread: conn34: url=jdbc:h2:/tmp/abcd user=SA
```

After:
```
testDeleteSelector():
<Nothing>

testReadSpecificDbResourceGroupCauseException():
11:51:45.625 [main] DEBUG org.javalite.activejdbc.DB -- Opened connection: conn37: url=jdbc:h2:/tmp/abcd user=SA
11:51:45.625 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Attached connection named: default: to current thread: conn37: url=jdbc:h2:/tmp/abcd user=SA. Extra info: jdbc:h2:/tmp/abcd
11:51:45.625 [main] INFO org.javalite.activejdbc.LazyList -- {"sql":"SELECT * FROM resource_groups","params":[],"cache":"hit"}
11:51:45.626 [main] DEBUG org.javalite.activejdbc.DB -- Closed connection: conn37: url=jdbc:h2:/tmp/abcd user=SA
11:51:45.626 [main] DEBUG org.javalite.activejdbc.ConnectionsAccess -- Detached connection named: default from current thread: conn37: url=jdbc:h2:/tmp/abcd user=SA

java.lang.AssertionError: 
Expecting code to raise a throwable.

	at io.trino.gateway.ha.router.TestSpecificDbResourceGroupsManager.testReadSpecificDbResourceGroupCauseException(TestSpecificDbResourceGroupsManager.java:73)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/Selectors.java [hunk 1]
```diff
@@ -13,85 +13,15 @@
...
-
-@BelongsTo(parent = ResourceGroups.class, foreignKeyName = "resource_group_id")
-@IdName("resource_group_id")
-@Table("selectors") // located in gateway-ha-persistence-*.sql
-@Cached
-public class Selectors
-        extends Model
+import org.jdbi.v3.core.mapper.reflect.ColumnName;
+
+public record Selectors(
```
Comment (mosabua) [inline gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/Selectors.java:18]: Dont we also want a  @JdbiConstructor in this class?

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                resource_group_id = :resourceGroupId
+                AND priority      = :priority
+                AND user_regex    IS NOT DISTINCT FROM :userRegex
+                AND source_regex  IS NOT DISTINCT FROM :sourceRegex
+                AND query_type    IS NOT DISTINCT FROM :queryType
+                AND client_tags   IS NOT DISTINCT FROM :clientTags
+                AND selector_resource_estimate IS NOT DISTINCT FROM :selectorResourceEstimate
+            LIMIT 1
+            """)
+    Selectors findFirst(@BindBean ResourceGroupsManager.SelectorsDetail selector);
```
Comment (mosabua) [inline gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java:48]: To be honest I dont like this pattern .. wouldnt it be better to have a unique id for each . and find by Id .. 

And yes .. I realize that would be a change of the data model and should probably be a separate PR..  just wanted to bring it up since this really should only return one record .. but nothing seems to really enforce that on the db level

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                selector_resource_estimate = :updatedSelector.selectorResourceEstimate
+            WHERE
+                resource_group_id = :selector.resourceGroupId
+                AND priority      = :selector.priority
+                AND user_regex    IS NOT DISTINCT FROM :selector.userRegex
+                AND source_regex  IS NOT DISTINCT FROM :selector.sourceRegex
+                AND query_type    IS NOT DISTINCT FROM :selector.queryType
+                AND client_tags   IS NOT DISTINCT FROM :selector.clientTags
+                AND selector_resource_estimate IS NOT DISTINCT FROM :selector.selectorResourceEstimate
+            LIMIT 1
```
Comment (mosabua) [inline gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java:75]: Same here .. we limit one and hope that we get the right record..

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            WHERE resource_group_id = :resourceGroupId
+            """)
+    List<Selectors> findByResourceGroupId(long resourceGroupId);
+
+    @SqlQuery("""
+            SELECT * FROM selectors
+            WHERE
+                resource_group_id = :resourceGroupId
+                AND priority      = :priority
+                AND user_regex    IS NOT DISTINCT FROM :userRegex
```
Comment (mosabua) [inline gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java:41]: Is that "IS NOT DISTINCT FROM" supported by all databases (postgresql, myself, and soon oracle)

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaResourceGroupsManager.java [hunk 5]
```diff
@@ -202,32 +184,13 @@ public List<SelectorsDetail> readSelector(long resourceGroupId,
...
-            if (model == null) {
-                Selectors.create(new Selectors(), updatedSelector);
-            }
-            else {
-                Selectors.update(model, updatedSelector);
-            }
+        SelectorsDao dao = getSelectorsDao(routingGroupDatabase);
+        Selectors model = dao.findFirst(selector);
+        if (model == null) {
+            dao.insert(updatedSelector);
```
Comment (mosabua) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaResourceGroupsManager.java:190]: Seems like a bad idea to silently do an insert instead of an update

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestSpecificDbResourceGroupsManager.java [hunk 6]
```diff
@@ -70,6 +70,9 @@ private void createResourceGroup()
     @Test
     public void testReadSpecificDbResourceGroupCauseException()
     {
+        // Cache could hit across different db url in activejdbc
+        // This manually clear the cache
+        org.javalite.activejdbc.cache.QueryCache.instance().purgeTableCache("resource_groups");
```
Comment (mosabua) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestSpecificDbResourceGroupsManager.java:75]: will get rid of this later I assume.

**Review by mosabua — COMMENTED** [review]
A couple of questions mostly. I am not too familiar with activejdbc or jdbi. The patterns used for modelling the db and the code seems to be a bit painful to me, but that might be my missing familiarity with it all. Anyway .. I kinda think we should get merge both of the jdbi refactor PRs and then look at the while persistence at once and figure out what we can improve in terms of the used SQL, the consistency of the different record classes and so on.

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java (hunk 2 above)
Comment (oneonestar) [reply gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java:48]: The `selectors` table is defined in Trino.
We have to update the DDL in Trino and carefully avoid breaking backward compatibility.
https://github.com/trinodb/trino/blob/53ad89113b7d1fdbafc6f89a0a75452c14300e1f/plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/ResourceGroupsDao.java#L71-L81

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java (hunk 4 above)
Comment (oneonestar) [reply gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java:41]: "IS NOT DISTINCT FROM" is defined in SQL:2003 standard. Trino supports it 10 years ago. PostgreSQL and MySQL supports it. Not sure about Oracle.

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java (hunk 3 above)
Comment (oneonestar) [reply gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java:75]: Same as above.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaResourceGroupsManager.java (hunk 5 above)
Comment (oneonestar) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaResourceGroupsManager.java:190]: This logic is same as before. I don't want to change the logic in code refactoring.

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestSpecificDbResourceGroupsManager.java (hunk 6 above)
Comment (oneonestar) [reply gateway-ha/src/test/java/io/trino/gateway/ha/router/TestSpecificDbResourceGroupsManager.java:75]: It will be gone in #287

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/Selectors.java (hunk 1 above)
Comment (oneonestar) [reply gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/Selectors.java:18]: https://jdbi.org/#_constructormapper
`@JdbiConstructor` is required when there are multiple constructors.

Comment (mosabua) [reply gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/Selectors.java:18]: oh .. thanks for explaining.

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java (hunk 2 above)
Comment (mosabua) [reply gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java:48]: Ok.. thats a separate large step .. definitely dont want to worry about that now.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaResourceGroupsManager.java (hunk 5 above)
Comment (mosabua) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaResourceGroupsManager.java:190]: yeah ... time for another refactor/thinking after this PR maybe.

**Review by mosabua — APPROVED** [review]
This looks good to me in terms of moving it over from ActiveJDBC to JDBI, and really that is the scope of this PR. I think we should ship this and the other refactor PR so we get a clean slate of JDBI only in the next release. Any concerns with merging both @ebyhr ?

File: gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java (hunk 3 above)
Comment (oneonestar) [reply gateway-ha/src/main/java/io/trino/gateway/ha/persistence/dao/SelectorsDao.java]: May bad. UPDATE statement shouldn't have LIMIT. Let me fix it.

**Comment by mosabua** [comment]
Merging now as agreed in the dev sync. If anything comes up we will address in follow up PRs.
